x/staking/keeper/determinstic_test.go

```diff
@@ -0,0 +1,945 @@
+	}
+
+	suite.setValidator(validator)
+	return validator
+}
+
+// getStaticValidator2 creates a validator with hard-coded values and sets to the state.
+func (suite *DeterministicTestSuite) getStaticValidator2() stakingtypes.Validator {
+	pubkey := ed25519.PubKey{Key: []byte{40, 249, 115, 32, 97, 18, 1, 1, 127, 255, 103, 13, 1, 34, 1, 24, 179, 242, 2, 151, 3, 34, 6, 1, 11, 0, 194, 202, 201, 77, 1, 167}}
+	pubkeyAny, err := codectypes.NewAnyWithValue(&pubkey)
+	suite.Require().NoError(err)
+
+	validator := stakingtypes.Validator{
+		OperatorAddress: validator2,
+		ConsensusPubkey: pubkeyAny,
+		Jailed:          true,
+		Status:          stakingtypes.Bonded,
+		Tokens:          sdk.NewInt(10012),
+		DelegatorShares: sdk.NewDecWithPrec(96, 2),
+		Description: stakingtypes.NewDescription(
+			"moniker",
+			"identity",
+			"website",
+			"securityContact",
+			"details",
+		),
+		UnbondingHeight: 100132,
+		UnbondingTime:   time.Date(2025, 10, 1, 0, 0, 0, 0, time.UTC),
+		Commission: stakingtypes.NewCommission(
+			sdk.NewDecWithPrec(15, 2),
+			sdk.NewDecWithPrec(59, 2),
+			sdk.NewDecWithPrec(51, 2),
+		),
+		MinSelfDelegation: sdk.NewInt(1),
+	}
+	suite.setValidator(validator)
+
+	return validator
+}
+
+// createDelegationAndDelegate funds the delegator account with a random delegation in range 100-1000 and delegates.
+func (suite *DeterministicTestSuite) createDelegationAndDelegate(t *rapid.T, delegator sdk.AccAddress, validator stakingtypes.Validator) (newShares math.LegacyDec, err error) {
+	amt := suite.stakingKeeper.TokensFromConsensusPower(suite.ctx, rapid.Int64Range(100, 1000).Draw(t, "amount"))
+	return suite.fundAccountAndDelegate(delegator, validator, amt)
+}
+
+// fundAccountAndDelegate funds the delegator account with the specified delegation and delegates.
+func (suite *DeterministicTestSuite) fundAccountAndDelegate(delegator sdk.AccAddress, validator stakingtypes.Validator, amt math.Int) (newShares math.LegacyDec, err error) {
+	coins := sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, amt))
+
+	suite.Require().NoError(suite.bankKeeper.MintCoins(suite.ctx, minttypes.ModuleName, coins))
+	banktestutil.FundAccount(suite.bankKeeper, suite.ctx, delegator, coins)
+
+	shares, err := suite.stakingKeeper.Delegate(suite.ctx, delegator, amt, stakingtypes.Unbonded, validator, true)
+	return shares, err
+}
+
+func (suite *DeterministicTestSuite) runValidatorIterations(
+	req *stakingtypes.QueryValidatorRequest,
+	prevValRes *stakingtypes.QueryValidatorResponse,
+	gasConsumed uint64,
+) {
+	for i := 0; i < 1000; i++ {
+		before := suite.ctx.GasMeter().GasConsumed()
+		res, err := suite.queryClient.Validator(suite.ctx, req)
+		suite.Require().Equal(suite.ctx.GasMeter().GasConsumed()-before, gasConsumed)
+
+		suite.Require().NoError(err)
+		suite.Require().Equal(res, prevValRes)
+	}
+}
+
+func (suite *DeterministicTestSuite) TestGRPCValidator() {
+	rapid.Check(suite.T(), func(t *rapid.T) {
+		suite.SetupTest() // reset
```

Comment:
What happens if we do `suite.SetupTest` outside of rapid.Check? I think it's a stronger test, because we test a bigger state